File: examples/tsj38-kotlin-parity/ts-app/src/service/order-service.ts
import type { OrderRepository } from "../repository/order-repository";
import type { SecurityPolicy } from "../security/security-policy";

export class OrderService {
  private readonly repository: OrderRepository;
  private readonly policy: SecurityPolicy;

  constructor(repository: OrderRepository, policy: SecurityPolicy) {
    this.repository = repository;
    this.policy = policy;
  }

  list(role: string) {
    if (!this.policy.canRead(role)) {
      throw "forbidden";
    }
    return this.repository.list();
  }

  create(role: string, customer: string, total: number) {
    if (!this.policy.canWrite(role)) {
      throw "forbidden";
    }
    return this.repository.save(customer, total);
  }
}

File: examples/tsj38-kotlin-parity/kotlin-app/src/main/kotlin/dev/tsj/reference/service/OrderService.kt
package dev.tsj.reference.service

import dev.tsj.reference.repository.OrderRepository
import dev.tsj.reference.security.SecurityPolicy

class OrderService(
    private val repository: OrderRepository,
    private val policy: SecurityPolicy
) {
    fun list(role: String): Any {
        if (!policy.canRead(role)) {
            throw IllegalStateException("forbidden")
        }
        return repository.list()
    }

    fun create(role: String, customer: String, total: Double): Any {
        if (!policy.canWrite(role)) {
            throw IllegalStateException("forbidden")
        }
        return repository.save(customer, total)
    }
}
